### PR TITLE
geo-layers: refactor tile-3d-layer

### PR DESCRIPTION
**Pull request by xintongxia**
#### Background
- loaders.gl v2.1 refactors change the tile loaders api

#### Change List
- Tile3DLayer API change

This PR depends on loaders.gl v2.1

### CODE REVIEW

File: docs/layers/tile-3d-layer.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Tile3DLayer (Experimental)
 
 The `Tile3DLayer` renders tileset data formatted according to the [3D Tiles Specification](https://www.opengeospatial.org/standards/3DTiles),
-which is supported by the [Tileset3DLoader](https://loaders.gl/modules/3d-tiles/docs/api-reference/tileset-3d-loader).
+which is supported by the [Tiles3DLoader](https://loaders.gl/modules/3d-tiles/docs/api-reference/tileset-3d-loader).
```
Comment (ibgreen) [inline docs/layers/tile-3d-layer.md]: Nit: If you have renamed the class, should the URL also change?

File: docs/layers/tile-3d-layer.md [hunk 2]
```diff
@@ -11,14 +11,23 @@ References
...
+      return null;
+    }
     const layer = new Tile3DLayer({
       id: 'tile-3d-layer',
-      data: '<path-to-your-tileset json file>',
+      data: 'https://api.cesium.com/v1/assets/49378',
+      loader: Tiles3DLoader,
+      // https://cesium.com/docs/rest-api/
+      loadOptions: {
+        headers: {Authentication: `Bear <ion-access-token>`}
```
Comment (ibgreen) [inline docs/layers/tile-3d-layer.md]: Nit: `Bearer`?

File: docs/layers/tile-3d-layer.md [hunk 3]
```diff
@@ -11,14 +11,23 @@ References
...
 import React, {Component} from 'react';
 import DeckGL from '@deck.gl/react';
+import {Tiles3DLoader} from '@loaders.gl/3d-tiles';
 import {Tile3DLayer} from '@deck.gl/geo-layers';
 
 export default class App extends Component {
-
   render() {
+    const {dataUrl, metadata} = this.state;
+    if (!dataUrl) {
```
Comment (ibgreen) [inline docs/layers/tile-3d-layer.md]: Is `dataUrl` and `metadata` used below?

File: docs/layers/tile-3d-layer.md [hunk 4]
```diff
@@ -100,11 +109,6 @@ This value is only applied when [tile format](https://github.com/AnalyticalGraph
 
 - A URL to fetch tiles entry point [Tileset JSON](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/specification#tileset-json) file or json object of tileset.
 
-##### `_ionAssetId` (Number|String, Optional)
-##### `_ionAccessToken` (String, Optional)
-
-- `_ionAssetId` and `_ionAccessToken` are used to fetch ion dataset. They are experimental properties, may change in next releases. 
-
 [Set up Ion account](https://cesium.com/docs/tutorials/getting-started/#your-first-app);
```
Comment (ibgreen) [inline docs/layers/tile-3d-layer.md]: Seems a little weird to have this line here when you cut the `ion` headers. Maybe place it somewhere else?

File: docs/layers/tile-3d-layer.md [hunk 5]
```diff
@@ -100,11 +109,6 @@ This value is only applied when [tile format](https://github.com/AnalyticalGraph
...
 
 ##### `loadOptions` (Object, Optional)
```
Comment (ibgreen) [inline docs/layers/tile-3d-layer.md]: I find this name a little confusing. I would have called this prop `tilesetOptions`. Not sure if it is worth renaming it now.

File: docs/layers/tile-3d-layer.md (hunk 3 above)
Comment (xintongxia) [reply docs/layers/tile-3d-layer.md]: removed

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 6]
```diff
@@ -251,8 +203,78 @@ export default class Tile3DLayer extends CompositeLayer {
...
+      }
+    );
+  }
+
+  _createSimpleMeshLayer(tileHeader) {
+    const content = tileHeader.content;
+    const {attributes, modelMatrix, cartographicOrigin, texture} = content;
+    const {normals, texCoords} = attributes;
+    const positions = new Float32Array(attributes.positions.value.length);
+    for (let i = 0; i < positions.length; i += 3) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: Why do you need to manually transform the positions? The layer’s modelMatrix should propagate in meter offsets mode.

File: docs/layers/tile-3d-layer.md [hunk 7]
```diff
@@ -171,6 +174,7 @@ The Tile3DLayer renders the following sublayers based on tile [format](https://g
 * `scenegraph` - a [ScenegraphLayer](/docs/layers/scenegraph-layer.md) rendering all the tiles with Batched 3D Model format (`b3dm`) or Instanced 3D Model format (`i3dm`).
   - `_lighting` is default to `pbr`.
 * `pointcloud` - a [PointCloudLayer](/docs/layers/point-cloud-layer.md) rendering all the tiles with Point Cloud format (`pnts`).
+* `simplemesh` - a [SimpleMeshLayer](/docs/layers/simple-mesh-layer.md) rendering all the tiles ESRI `MeshPyramids` data.
```
Comment (Pessimistress) [inline docs/layers/tile-3d-layer.md]: Nit: just call it `mesh`

File: docs/upgrade-guide.md [hunk 8]
```diff
@@ -1,5 +1,13 @@
 # Upgrade Guide
 
+## Upgrading from deck.gl v8.0 to v8.1
+
+### Breaking Changes
+
+##### Tile3DLayer
+
+`_ionAccessId` and `_ionAccesToken` are removed. To render an ion dataset with `Tile3DLayer`, pass the ion dataset url to prop `data`, and `loadOptions.headers` with Cesium authentication token.
```
Comment (Pessimistress) [inline docs/upgrade-guide.md]: The layer will attempt to download `data` and parse with registered loaders. Use a different prop name if that is not the intention.

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 9]
```diff
@@ -67,84 +64,68 @@ export default class Tile3DLayer extends CompositeLayer {
...
+    const tileset3d = new Tileset3D(tilesetJson, {
+      onTileLoad: this._onTileLoad.bind(this),
       onTileUnload: this.props.onTileUnload,
       onTileLoadFail: this.props.onTileError,
-      // TODO: explicit passing should not be needed, registerLoaders should suffice
-      fetchOptions,
-      ...ionMetadata,
-      ...loadOptions
+      fetchOptions: loadOptions
     });
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: So users will not be able to set any tiles loader options other than `fetch`?

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 10]
```diff
@@ -67,84 +64,68 @@ export default class Tile3DLayer extends CompositeLayer {
...
-    const loadOptions = this.getLoadOptions();
-
-    const tileset3d = new Tileset3D(tilesetJson, tilesetUrl, {
-      onTileLoad: tileHeader => {
-        this.props.onTileLoad(tileHeader);
-        this._updateTileset(tileset3d);
-      },
+  async _loadTileset(tilesetUrl) {
+    const {loader, loadOptions} = this.props;
+    const tilesetJson = await load(tilesetUrl, loader, loadOptions);
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: What if loader is not supplied? Add helpful warning to help users cope with the breaking change.

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 11]
```diff
@@ -67,84 +64,68 @@ export default class Tile3DLayer extends CompositeLayer {
...
   }
 
-  async _loadTilesetFromIon(ionAccessToken, ionAssetId) {
-    const ionMetadata = await _getIonTilesetMetadata(ionAccessToken, ionAssetId);
-    const {url, headers} = ionMetadata;
-    await this._loadTileset(url, {headers}, ionMetadata);
+  _onTileLoad(tileHeader) {
+    this.props.onTileLoad(tileHeader);
+    this._updateTileset(this.state.tileset3d);
+    this.setNeedsUpdate();
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: Why is this needed? What happened to `frameNumber`?

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 12]
```diff
@@ -251,8 +203,78 @@ export default class Tile3DLayer extends CompositeLayer {
...
+    );
+  }
+
   renderLayers() {
-    return this.state.layers;
+    const {tileset3d, layerMap} = this.state;
+    if (!tileset3d) {
+      return null;
+    }
+    this._updateLayerMap();
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: Too many intermediate arrays are created here. Directly call `tileset3d.tiles.map` similar to the 2d tiles layer.

File: modules/geo-layers/package.json [hunk 13]
```diff
@@ -30,8 +30,10 @@
     "prepublishOnly": "npm run build-bundle && npm run build-bundle -- --env.dev"
   },
   "dependencies": {
-    "@loaders.gl/3d-tiles": "^2.1.0-alpha.4",
-    "@loaders.gl/terrain": "^2.1.0-alpha.4",
+    "@loaders.gl/3d-tiles": "^2.1.0",
```
Comment (Pessimistress) [inline modules/geo-layers/package.json]: Use a beta release so that we can test.

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js (hunk 11 above)
Comment (xintongxia) [reply modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: > Why is this needed?
After `this._updateTileset`, the `renderLayers` should be called, which it isn't if I don't manually call `this.setNeedsUpdate()`.

> What happened to frameNumber?
`frameNumber` is not needed to decide if the tile should be selected for rendering, `tile.selected` is enough.

Comment (Pessimistress) [reply modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: `frameNumber` is used to inform the layer whether the selected tiles have changed. We should not have to rerender on every viewport change.

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js (hunk 6 above)
Comment (xintongxia) [reply modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: positions are `Float64`, can I construct a geometry with `Float64Array`?

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js (hunk 9 above)
Comment (xintongxia) [reply modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: I'll update this, should allow user to pass in other loadOptions

File: docs/upgrade-guide.md (hunk 8 above)
Comment (xintongxia) [reply docs/upgrade-guide.md]: `data` of `Tile3DLayer` is the url to fetch entry point tileset file.

I am proposing the following change:

``` 
    const layer = new Tile3DLayer({
      id: 'tile-3d-layer',	      id: 'tile-3d-layer',
      data: '<path-to-your-tileset json file>',	      data: 'https://api.cesium.com/v1/assets/49378',
      loader: Tiles3DLoader,
      // https://cesium.com/docs/rest-api/
      loadOptions: {
        headers: {Authentication: `Bearer <your_ion_access_token>`} 
      },
      ...
    })
```

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js (hunk 6 above)
Comment (Pessimistress) [reply modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: Casting it to Float32Array is fine. Why do you need to apply the modelMatrix?

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 14]
```diff
@@ -251,8 +164,111 @@ export default class Tile3DLayer extends CompositeLayer {
...
+    return new SubLayerClass(
+      {
+        _lighting: 'pbr'
+      },
+      this.getSubLayerProps({
+        id: 'scenegraph'
+      }),
+      {
+        id: `${this.id}-scenegraph-${tileHeader.id}`,
+        // Fix for ScenegraphLayer.modelMatrix, under flag in deck 7.3 to avoid breaking existing code
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: This comment no longer applies

File: modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js [hunk 15]
```diff
@@ -251,8 +164,111 @@ export default class Tile3DLayer extends CompositeLayer {
...
   renderLayers() {
-    return this.state.layers;
+    const {tileset3d, layerMap} = this.state;
+    if (!tileset3d) {
+      return null;
+    }
+
+    return tileset3d.tiles
+      .map(tile => {
+        if (tile.contentUnloaded) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/tile-3d-layer/tile-3d-layer.js]: Do this in `onTileUnload`

File: docs/upgrade-guide.md (hunk 8 above)
Comment (ibgreen) [reply docs/upgrade-guide.md]: To get the ion-access-token suitable for use as `Bearer`, isn't it necessary to first do a couple of ping-pongs with the Cesium API? We had a function for that...

If this works with the tokens users normally use then things are fine.

But if it requires calling a separate function to get the required Authentication header token, then it will no longer be possible to show 3D tilesets in the playground.

Comment (xintongxia) [reply docs/upgrade-guide.md]: I checked the [code](https://github.com/uber-web/loaders.gl/blob/master/modules/3d-tiles/src/lib/ion/ion.js), also Cesium [website](https://cesium.com/docs/tutorials/rest-api/#authentication), to access the tiles, it only needs the auth header.

Comment (ibgreen) [reply docs/upgrade-guide.md]: >  it only needs the auth header.

Yes, I believe that is correct, as long as you have the auth header you can access the tiles. My question is how does the application get the auth header in the first place. I seem to recall having to do a ping-pong with the Cesium server to get that header, feeding the ion tokens and ultimately getting the auth header back.

See https://github.com/uber-web/loaders.gl/blob/9c0a8c74dba44e393eb671113b5f3e96768327a8/modules/3d-tiles/src/lib/ion/ion.js

Comment (xintongxia) [reply docs/upgrade-guide.md]: - If the user provides both the assetId and access token, then only needs pass in header.
- If user only provides access token, then the ion util function will try to fetch the all the assets associated with that token and find one asset id.

The current API can cover case 1, for case 2, I am thinking I could provide doc/example of using ion utils to get the asset url pass in `Tile3DLayer`.

The rest of ion util is for getting ion dataset metadata. I feel that is an application level of detail. I [staged](http://xintongxia.github.io/) my changes and tested with examples, it didn't have auth issues to request the tiles.

Comment (ibgreen) [reply docs/upgrade-guide.md]: > am thinking I could provide doc/example of using ion utils to get the asset url pass in Tile3DLayer. I feel that is an application level of detail. 

@xintongxia Thanks for explaining. I am not disagreeing with the logic of providing the ion utils separately. 

My only real concern is what happens to the [JSON playground examples](https://github.com/uber/deck.gl/blob/master/examples/playground/json-examples/3d-tiles-st-helens.json#L25). They will no longer be able to reference ION tokens, right? Is your plan to list `headers: {Authentication: `"... accessToken"}` in the JSON? 

I don't think we are supposed to store the access tokens returned from Cesium ION directly in our github repo, and also I assume those tokens can be rotated by Cesium at any time so even it works during testing it may fail a week later.

One approach could be that you include any required changes to the playground apps  this PR - either make sure the playground examples still work without `ionAccessToken` prop, or conversely, deleted those examples, if your decision is that we now require apps to call the ion utility functions to use Tile3DLayer with Cesium ION.

Making sure that we agree that this PR may leave the 3D tiles playground examples broken is really my only concern. I am fine with however you decide to resolve that.
